Show border, addButton and Quanta Toolbar over Block when selected

MAX-786 opened the pull request:
Fixes #24, fixes #16, fixes #25, fixes #27, fixes #26

#### Features:

- All these 3 things (border, add & delete btn) are injected by Bridge(hydra.js) only when it is loaded inside Hydra and it is in edit mode. 
- Click on block and border and buttons will appear.
- Click on add button and it will add a slate (default) block below the selected block and the border will shift to that new block (now using mutation observer).
- You can also edit text of slate blocks from sidebar and it is sent to iframe in real time.
- Color and design are subjected to change as per review.

#### Goals in order of execution:

- [x] Border is rendered by hydra.js and it is same for all block types. (as discussed this is enough for the scope of this PR)
- [x] Add button and delete button injected by hydra.js.
- [x] Add button adds a text block block and border is moved to new block.
- [x] Now Add button opens up a BlockChooser Menu and it is rendered by adminUI over the top of iframe.
- [x]  Clicking on block will show up quanta toolbar injected by hydra.js and will have drag btn (not functional) and three dot menu which will include Remove btn. (As for now it's the same for mobile)
- [x] Placing the components as per quanta design .

#### Changes to previous implementations:

- Instead of using redux state, Now using `onSelectBlock()` method to update sidebar when block is selected in Iframe.
- Using `props` from Form.jsx to `addBlock`  & `deleteBlock` from the formData.

#### Current problems:

- (please let us know if you find any)

#### Current Design:


https://github.com/collective/volto-hydra/assets/99530996/b76aa98c-b65c-4c18-a008-58360f4865df

djay commented:
as mentioned in the meeting. The add popup doesn't have to match the quanta styling. 
The quanta toolbar would just have the drag icon (not working/disabled) and the ... menu.
There are mobile aspects to how the quanta toolbar and add menu are placed. but maybe could split those out

JeffersonBledsoe commented:
@MAX-786 I haven't figured out why yet, but the issue with the `onDeleteBlock` call not working properly is that the `this.props.globalData` isn't being updated. So when the component updates, it replaces the existing form data with the stale global data, causing it to bring the old block back. See https://github.com/collective/volto-hydra/blob/b8aa970c0aec766de59af8fc937d537b5a518e83/packages/volto-hydra/src/customizations/components/manage/Form/Form.jsx#L286-L293 for where this happens. 
`this.props.globalData` is updated by calling `this.props.setFormData`. So either we've removed a call to this, or something we have changed is causing this call to get out of sync with the actual form data

MAX-786 commented:
@JeffersonBledsoe Thanks and Yepp, I fixed it!!! Well globalData was getting updated but not inside the `onDeleteBlock()` method defined in Iframe component. Read Below for more details!
The problem was in the Iframe component, here See:
https://github.com/collective/volto-hydra/blob/4efcbf572978913a01aaab7392bdba8f71d51c31/packages/volto-hydra/src/components/Iframe/View.jsx#L75-L88

You can see `onDeleteBlock` was defined outside of the useEffect, which means it captured and used the `properties` state from the time it was defined. If `properties` updated, `onDeleteBlock` was still using the old `properties` state, leading to stale or incorrect state usage.

But when i moved the `onDeleteBlock` function inside the useEffect, it gets redefined every time the useEffect runs. This ensures that it always captures the latest state and props, including the updated `properties`.

I hope this is the correct way to do this. :)

MAX-786 commented:
I will be pushing the code now, with addblock feature using the BlockChooser form and deleteBlock btn still at the right side of the block (since quanta toolbar is now the next target of this PR and i'll be moving delete btn in quanta toolbar)

JeffersonBledsoe commented:
> @JeffersonBledsoe Thanks and Yepp, I fixed it!!! Well globalData was getting updated but not inside the `onDeleteBlock()` method defined in Iframe component. Read Below for more details!
> 
> The problem was in the Iframe component, here See:
> 
> https://github.com/collective/volto-hydra/blob/4efcbf572978913a01aaab7392bdba8f71d51c31/packages/volto-hydra/src/components/Iframe/View.jsx#L75-L88
> 
> 
> 
> You can see `onDeleteBlock` was defined outside of the useEffect, which means it captured and used the `properties` state from the time it was defined. If `properties` updated, `onDeleteBlock` was still using the old `properties` state, leading to stale or incorrect state usage.
> 
> 
> 
> But when i moved the `onDeleteBlock` function inside the useEffect, it gets redefined every time the useEffect runs. This ensures that it always captures the latest state and props, including the updated `properties`.
> 
> 
> 
> I hope this is the correct way to do this. :)

Good catch on the stale closure! I think the more "React" fix would be to set properties as one of the dependencies within the use effect, but I'd need to look more closely when I get back. Feel free to continue with this approach for now though if it works so we can continue the project :)

JeffersonBledsoe commented:
I haven't looked properly, but go for it if the functionality is there as we can improve it later

MAX-786 commented:
Yeppp same thought :)
Then I'll wait for Dylan's review.

djay commented on README.md:
```diff
@@ -260,9 +260,13 @@ onEditChange(initialData, handleEditChange);
+
+- You can click on '+' Icon (appears at the bottom-right of the container in which you added `data-bloc-uid="<<BLOCK_UID>>>"` attribute) to add a block below the current block by choosing a type from BlockChooser popup.
+- You can click on three dots icon on Quanta toolbar (appears at the top-left) and it will open up a dropdown menu, you can click on 'Remove' to delete the current block.
+- Settings option is yet to be implemented (TODO)
```
And a new ticket for stuff to be done later and include that ticket url on the TODO

djay commented on README.md:
```diff
@@ -260,9 +260,13 @@ onEditChange(initialData, handleEditChange);
 
 ### Level 4: Enable Managing Blocks directly on your frontend ([TODO](https://github.com/collective/volto-hydra/issues/4))
```
Since the main part of the this stage is done you can remove this TODO. extra functions to be added later will have their own TODO links below

djay commented on README.md:
```diff
@@ -260,9 +260,13 @@ onEditChange(initialData, handleEditChange);
+- You can click on '+' Icon (appears at the bottom-right of the container in which you added `data-bloc-uid="<<BLOCK_UID>>>"` attribute) to add a block below the current block by choosing a type from BlockChooser popup.
+- You can click on three dots icon on Quanta toolbar (appears at the top-left) and it will open up a dropdown menu, you can click on 'Remove' to delete the current block.
+- Settings option is yet to be implemented (TODO)
 - drag and drop blocks ([TODO](https://github.com/collective/volto-hydra/issues/65))
```
if possible it would be good to show the drag icon disabled just to indicate it doesn't work yet

djay commented on README.md:
```diff
@@ -260,9 +260,13 @@ onEditChange(initialData, handleEditChange);
-- remove blocks ([TODO](https://github.com/collective/volto-hydra/issues/26))
+If you completed level 2 & 3 (made blocks clickable and enabled live updates) then the editor will automatically gain the management of blocks on the frontend using the quanta toolbar.
+
+With Quanta toobar, you can use following features:
```
if possible it would be good to add a shadow to the bar so it stands out more. See what you can take from the quanta design to make it show up better against white.

MAX-786 replied:
@djay  Will something like this works?

![Screenshot from 2024-06-28 08-37-06](https://github.com/collective/volto-hydra/assets/99530996/03675bc2-9b2d-4a31-b8b5-54d3f447b49f)

djay reviewed (changes requested):
It's looking good. just a few minor UI and readme tweaks

djay commented:
@MAX-786 he is also using grey on the drag icon to help the toolbar stand out. Would be good to add that.

![image](https://github.com/collective/volto-hydra/assets/41700/352829a9-fc5d-4ea7-8e32-d84a047324f6)

djay commented on README.md:
```diff
@@ -258,11 +258,15 @@ function handleEditChange(updatedData) {
+
+- You can click on '+' Icon (appears at the bottom-right of the container in which you added `data-bloc-uid="<<BLOCK_UID>>>"` attribute) to add a block below the current block by choosing a type from BlockChooser popup.
+- You can click on three dots icon on Quanta toolbar (appears at the top-left) and it will open up a dropdown menu, you can click on 'Remove' to delete the current block.
+- Settings option is yet to be implemented [TODO](https://github.com/collective/volto-hydra/issues/14)
```
```suggestion
- Settings option is yet to be implemented [TODO](https://github.com/collective/volto-hydra/issues/81)
```

djay commented on README.md:
```diff
@@ -258,11 +258,15 @@ function handleEditChange(updatedData) {
+
+With Quanta toobar, you can use following features:
+
+- You can click on '+' Icon (appears at the bottom-right of the container in which you added `data-bloc-uid="<<BLOCK_UID>>>"` attribute) to add a block below the current block by choosing a type from BlockChooser popup.
```
if you look at the screen shot I think he also puts a shadow on the add menu and moves it out from the right and bottom edges somewhat to make it stand out a bit more

djay replied:
I'll split out the behavour for mobile into other cards

MAX-786 replied:
> if you look at the screen shot I think he also puts a shadow on the add menu and moves it out from the right and bottom edges somewhat to make it stand out a bit more

You mean BlockChooser popup? if yes, then it already got a decent box-shadow tho.. we can try out more if you want to check it out,
and if you mean the 'plus' icon, then it already a nice blue border keeping it distinctive.

djay replied:
yes blockchooser. if its got a shadow then maybe it should needs to be moved out from the edges of the iframe a bit more.

MAX-786 replied:
Will this much works?

![Screenshot from 2024-06-29 08-09-15](https://github.com/collective/volto-hydra/assets/99530996/d017d5d4-0f99-4f47-b43a-64ea21dcb9f6)

djay commented on packages/hydra-js/hydra.js:
```diff
@@ -294,7 +295,7 @@ class Bridge {
       }
       .volto-hydra-drag-button {
-        cursor: grab;
+        cursor: default;
```
give a grey background like the design

djay commented:
@MAX-786 looks good. We can merge it I think

MAX-786 commented:
Merging and closing it!
